streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java

```diff
@@ -0,0 +1,128 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The time-based window specifications used for aggregations.
+ */
+public class TimeWindows extends Windows<TimeWindow> {
+
+    /**
+     * The size of the window, i.e. how long a window lasts.
+     * The window size's effective time unit is determined by the semantics of the topology's
+     * configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     */
+    public final long size;
+
+    /**
+     * The size of the window's hop, i.e. by how much a window moves forward relative to the previous one.
+     * The hop size's effective time unit is determined by the semantics of the topology's
+     * configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     */
+    public final long hop;
+
+    private TimeWindows(String name, long size, long hop) {
+        super(name);
+        if (size <= 0) {
+            throw new IllegalArgumentException("size must be > 0 (you provided " + size + ")");
+        }
+        this.size = size;
+        if (!(0 < hop && hop <= size)) {
+            throw new IllegalArgumentException(String.format("hop (%d) must lie within interval (0, %d]", hop, size));
+        }
+        this.hop = hop;
+    }
+
+    /**
+     * Returns a window definition with the given window size, and with the hop size being equal to
+     * the window size.  Think: [N * size, N * size + size), with N denoting the N-th window.
+     *
+     * This provides the semantics of tumbling windows, which are fixed-sized, gap-less,
+     * non-overlapping windows.  Tumbling windows are a specialization of hopping windows.
+     *
+     * @param name The name of the window.  Must not be null or empty.
+     * @param size The size of the window, with the requirement that size &gt; 0.
+     *             The window size's effective time unit is determined by the semantics of the
+     *             topology's configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     * @return a new window definition
+     */
+    public static TimeWindows of(String name, long size) {
+        return new TimeWindows(name, size, size);
+    }
+
+    /**
+     * Returns a window definition with the original size, but shift ("hop") the window by the given
+     * hop size, which specifies by how much a window moves forward relative to the previous one.
+     * Think: [N * hop, N * hop + size), with N denoting the N-th window.
+     *
+     * This provides the semantics of hopping windows, which are fixed-sized, overlapping windows.
+     *
+     * @param hop The hop size of the window, with the requirement that 0 &lt; hop &le; size.
+     *            The hop size's effective time unit is determined by the semantics of the
+     *            topology's configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     * @return a new window definition
+     */
+    public TimeWindows shiftedBy(long hop) {
```

Comment:
I updated the code by renaming shiftedBy/hop to advanceBy/advance.  See the latest PR for details.